content/post/104-dht-deep-dive.md

```diff
@@ -0,0 +1,154 @@
+
+A routing table refresh works as follows:
+
+1. Go through all the buckets, from bucket 0 (the one with peers that are on a different half of the network than we are) up until the highest bucket we have that contains a peer in it (we cap this at bucket 15 due to implementation concerns). For each bucket select a random address in Kademlia space that could fit in that bucket (e.g., in selecting a random peer between 512 and 1024 we selected 678 even though that peer likely does not exist in the network) and do a lookup to find the `K` closest peers to that random address. This will ensure that we will have filled up each bucket with as many peers as will fit.
+2. We also search for ourselves in the network (i.e., bucket 255), just in case the network size and distribution are such that the first 15 buckets do not suffice for us to learn about the `K` peers closest to us.
+
+## Lookup Algorithm
+
+The lookup algorithm answers the question “What are the `K` closest peers to `X`?”. Our implementation of the Kademlia lookup algorithm goes like this:
+
+1. Load the `K` closest peers to `X` from our routing table into the query queue
+2. Allowing up to Alpha concurrent queries (Alpha in go-ipfs is 10, but is an implementation parameter not inherent in the network itself) grab the peer closest to `X` and ask them “who are the `K` closest peers to `X`?”
+3. When a query to a peer completes add those results to the query queue, and pull the next peer off the queue and query them
+4. The query terminates whenever the closest Beta peers to `X` have been successfully queried (i.e. no dial timeouts, errors, etc.).
+    - Note: Beta is a network wide parameter designed to give some resiliency to the network, it is set to 3 for IPFS.
+5. After the query is done take the `K` closest peers that we have not already failed at querying (i.e. we have heard from them or they are still in our queue) and return them
+    - Note: for some API compatibility reasons go-ipfs also ensures that we have actually sent queries to all of the top `K` peers
+
+In the routing table section, we mentioned that we evict peers that have not been “recently useful” to us if we find a new peer that could take their spot in the bucket.
+We define “recently” as the time period during which they are probabilistically expected to have been utilized in a refresh. That value is `Log(1/K) * Log(1 - Alpha/K) * refreshPeriod`, where Alpha is the number of peers dialed that can be simultaneously queried. Additionally, we define “useful” as responding within 2x the time it takes any other peer from our routing table to respond to us (this biases against peers that are slow, overloaded, unreliable, or have bad network connectivity to us). This definition of useful will likely change as we gather more information about the dynamics of the network and investigate relevant threat models.
```

Comment:
@petar If we change the corresponding section above then maybe something like the below (especially given that we have a different model, i.e. replaceable peers are only ones that we found during bootstrapping, scheduled for the next release)

```suggestion
In the routing table section, we mentioned that we evict peers that are "replaceable" if we find a new peer that could take their spot in the bucket.
In v0.5.0 we define a peer as "replaceable" if they have not been "useful" to us within the time period during which they are probabilistically expected to have been utilized in a refresh. That value is `Log(1/K) * Log(1 - Alpha/K) * refreshPeriod`, where Alpha is the number of peers dialed that can be simultaneously queried. Additionally, we define “useful” as responding within 2x the time it takes any other peer from our routing table to respond to us (this biases against peers that are slow, overloaded, unreliable, or have bad network connectivity to us). The definitions of replaceable and useful peers are likely to change as we gather more information about the dynamics of the network and investigate relevant threat models.
```

Reply:
This looks fine. It introduces "replaceable" to mean "not recently useful". And the definition of "recently" and "useful" is unchanged.